src/Nethermind/Nethermind.Synchronization/Peers/SyncPeerPool.cs

```diff
@@ -584,93 +592,257 @@ private async Task ExecuteRefreshTask(RefreshTotalDiffTask refreshTotalDiffTask,
-                        linkedSource.Dispose();
-                        delaySource.Dispose();
+                        headBlockHeader = headAndParentHeaders[1];
                     }
-                }, token);
+                    headParentBlockHeader = headAndParentHeaders[0];
+                }
+                else if (headAndParentHeaders.Length > 0)
+                {
+                    if (_logger.IsTrace) _logger.Trace($"PeerRefreshForFCU unexpected response length when fetching header: {headAndParentHeaders.Length}");
+                }
+
+                finalizedBlockHeader = await getFinalizedHeaderTask;
+            }
+            catch (TaskCanceledException exception)
+            {
+                if (_logger.IsTrace)
+                    _logger.Trace(
+                        $"PeerRefreshForFCU canceled for node: {syncPeer.Node:c}{Environment.NewLine}{exception}");
+                _stats.ReportSyncEvent(syncPeer.Node,
+                    syncPeer.IsInitialized
+                        ? NodeStatsEventType.SyncCancelled
+                        : NodeStatsEventType.SyncInitCancelled);
+                token.ThrowIfCancellationRequested();
+                return;
+            }
+            catch (Exception exception)
+            {
+                if (_logger.IsTrace)
+                    _logger.Trace($"PeerRefreshForFCU failed for node: {syncPeer.Node:c}{Environment.NewLine}{exception}");
+                _stats.ReportSyncEvent(syncPeer.Node,
+                    syncPeer.IsInitialized ? NodeStatsEventType.SyncFailed : NodeStatsEventType.SyncInitFailed);
+                // TODO: how to know if exception is transient
+                syncPeer.Disconnect(DisconnectReason.DisconnectRequested, "refresh peer info fault - faulted");
+                return;
+            }
+
+            if (_logger.IsTrace)
+            {
+                _logger.Trace($"PeerRefreshForFCU received block info from {syncPeer.Node:c}");
+                _logger.Trace($"PeerRefreshForFCU headHeader: {headBlockHeader}");
+                _logger.Trace($"PeerRefreshForFCU headParentHeader: {headParentBlockHeader}");
+                _logger.Trace($"PeerRefreshForFCU finalizedBlockHeader: {finalizedBlockHeader}");
+            }
+
+            if (finalizedBlockhash != Keccak.Zero && finalizedBlockHeader == null)
+            {
+                if (_logger.IsTrace)
+                    _logger.Trace($"PeerRefreshForFCU failed for node: {syncPeer.Node:c}{Environment.NewLine} - Finalized block header not found");
+                _stats.ReportSyncEvent(syncPeer.Node,
+                    syncPeer.IsInitialized ? NodeStatsEventType.SyncFailed : NodeStatsEventType.SyncInitFailed);
+                syncPeer.Disconnect(DisconnectReason.DisconnectRequested, "refresh peer info fault - no finalized block header");
+                return;
+            }
+
+            List<BlockHeader> headersToCheck = new [] {
+                headBlockHeader,
```

Comment:
@LukaszRozmej @asdacap do you guys think that we should ask for headBlock and parentHeadBlock? What if we ask only for parentHeadBlock?

Reply:
because I don't see value and we're adding more traffic to the network